### PR TITLE
[Wallet] Restructuring Wallet Analytics

### PR DESCRIPTION
**Pull request by tarikbellamine**
### Description
Restructured and renamed the Valora wallet analytics. Changes include:
- Organized analytics events by category and introduced type checking for parameters
- Refactored all analytics parameters to be primitive data types to avoid pipeline or database errors
- Created a sessionId property and attach it and user address on each logged event
- Deleted unused or redundant events and parameters
- Consolidated all analytics into the `mobile` directory (deleted parent component in `react-components`)
- Removed subevent logging
- Renaming `CeloAnalytics` to `ValoraAnalytics`

To cover in next PR:
- Standardize event naming
- Further consolidate events
- Audit against [analytics doc](https://docs.google.com/document/d/1OGNJBijiWp7bWc2I0p8yyJ72EMCj9pNRT1VWxJJu5JY/edit?pli=1#heading=h.wip61atcgd2c)
- Introduce new events to fill gaps
- Update documentation

### Other changes
Fixed a couple of events that were broken.

### Tested
Yes

### Related issues
- Part of #4008

### Backwards compatibility
Yes

### CODE REVIEW

File: packages/mobile/src/identity/privateHashing.ts
Comment (tarikbellamine) [inline packages/mobile/src/identity/privateHashing.ts]: @jmrossy does this look right to you? i don't think we were distinguishing between a skip and fail in the previous iteration

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (tarikbellamine) [inline packages/mobile/src/analytics/ValoraAnalytics.ts:73]: Is there a better way to generate a session identifier?

File: packages/mobile/src/identity/privateHashing.ts
Comment (jmrossy) [reply packages/mobile/src/identity/privateHashing.ts]: Looks okay. It might be more natural to just `reject('skipped')` and then compare the error to 'skipped' directly but this is fine if you've already tested this.

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [inline packages/mobile/src/analytics/ValoraAnalytics.ts]: ditto

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [inline packages/mobile/src/analytics/ValoraAnalytics.ts:19]: Can we remove this? I haven't had time to do the PII purge yet but this should probably go

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [reply packages/mobile/src/analytics/ValoraAnalytics.ts:73]: Does segment and/or firebase make one for us already?

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [inline packages/mobile/src/analytics/ValoraAnalytics.ts]: I think it's generally not advisable to due async / fail-able actions in constructors. You could instead move this to an async `initi` method

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [inline packages/mobile/src/analytics/ValoraAnalytics.ts]: @nityas how would you feel about leaving out the user address. Given the growing scope of analytics, I'm not sure how ethical it is to tie all this activity to particular addresses

**Review by jmrossy — COMMENTED** [review]
Mostly LGTM, just a few comments below

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (nityas) [reply packages/mobile/src/analytics/ValoraAnalytics.ts]: hm...on ethics I lean on the ability of any user to opt out of analytics at any time(which we mention in onboarding), which is far more than most consumer apps (and all comparables) offer. In general, we could always tie together events implicitly through anonymous IDs, so I see this as a way to make sure our data is accurate (this also sets us up better to accurately bulk delete user data if a request is made, which is a feature I could see us adding in the future)

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (tarikbellamine) [reply packages/mobile/src/analytics/ValoraAnalytics.ts:73]: I would think so but we haven't seen it come through in the data. At any rate, I think it's good to have universal identifiers that we have granular control over and aren't tied to any specific database or packages so the data remains easily portable.

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (tarikbellamine) [reply packages/mobile/src/analytics/ValoraAnalytics.ts]: @nityas are you good with cutting this? I agree that the probability of it containing PII is pretty high.

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (nityas) [reply packages/mobile/src/analytics/ValoraAnalytics.ts:19]: carrier is really helpful for understanding verification issues...

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (nityas) [reply packages/mobile/src/analytics/ValoraAnalytics.ts]: +1 to removing this

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (nityas) [reply packages/mobile/src/analytics/ValoraAnalytics.ts:19]: Feature request for analytics: ability to log encrypted fields & decrypt on-demand while querying, such that risk is minimized if analytics infrastructure is compromised

Comment (tarikbellamine) [reply packages/mobile/src/analytics/ValoraAnalytics.ts:19]: I'm certainly open to it but don't feel strongly either way.

Most countries only have a handful of carriers, which can be easily implied by country codes. I can see having the granularity of carrier be useful for analytics around verification but you are much closer to that than I am so I'd defer to your assessment of the tradeoffs.

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (tarikbellamine) [reply packages/mobile/src/analytics/ValoraAnalytics.ts]: Are there any specific data points you're hesitant to tie to address or is this more of a general concern?

I think no collection of PII, an easily accessible opt-out of all analytics, and clear documentation on what's being collected sets a sufficiently high bar.

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (tarikbellamine) [reply packages/mobile/src/analytics/ValoraAnalytics.ts]: Makes sense.

I'm not very familiar with this design pattern. Do you mind taking a look at how I've done it in my latest "PR feedback" commit?

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (annakaz) [reply packages/mobile/src/analytics/ValoraAnalytics.ts]: @jmrossy because we're tracking transaction hashes, we can identify the user address from the hash sent from a particular wallet. I agree that we should not log any PII (phone number, name, etc) but would argue that the address falls more in the category of on-chain data that is acceptable to track

File: packages/mobile/src/alert/actions.ts
Comment (annakaz) [inline packages/mobile/src/alert/actions.ts:69]: Any reason to remove this? Error_fallback was helping to catch missing errors- for example if verification failed we could see a more detailed error message than the one showed to users

File: packages/mobile/src/app/ErrorBoundary.tsx
Comment (annakaz) [inline packages/mobile/src/app/ErrorBoundary.tsx:27]: Could be useful to keep the device info for errors

File: packages/mobile/src/app/saga.ts
Comment (annakaz) [inline packages/mobile/src/app/saga.ts]: Add TODO?

File: packages/mobile/src/geth/geth.ts
Comment (annakaz) [inline packages/mobile/src/geth/geth.ts:322]: 👍  Thanks for removing the phone number here

File: packages/mobile/src/geth/saga.ts
Comment (annakaz) [inline packages/mobile/src/geth/saga.ts]: Looks like `result` is just of type GethInitOutcomes and can be passed without .message

File: packages/mobile/src/identity/verification.ts
Comment (annakaz) [inline packages/mobile/src/identity/verification.ts:76]: Generally opposed to client side timing metrics but agree with the trade off to have these easily accessible

File: packages/mobile/src/identity/verification.ts
Comment (annakaz) [inline packages/mobile/src/identity/verification.ts:89]: Do these durations end up passed as a string? Not sure how the pipeline handles date objects

File: packages/mobile/src/invite/saga.ts
Comment (annakaz) [inline packages/mobile/src/invite/saga.ts]: In this case the invite is empty, not the escrow account. There are two accounts used for invite codes: a temp account to pay for verification and some escrow amount sent after verification

File: packages/mobile/src/onboarding/registration/JoinCelo.tsx
Comment (annakaz) [inline packages/mobile/src/onboarding/registration/JoinCelo.tsx]: nit: typo "an"

File: packages/mobile/src/tokens/saga.ts
Comment (annakaz) [inline packages/mobile/src/tokens/saga.ts:80]: 👍  thanks for fixing

**Review by annakaz — APPROVED** [review]
LGTM overall, a few comments below. See @jmrossy suggestion on async constructors too

File: packages/mobile/src/identity/verification.ts
Comment (aaronmgdr) [reply packages/mobile/src/identity/verification.ts:89]: `Date.now` returns a number

Comment (tarikbellamine) [reply packages/mobile/src/identity/verification.ts:89]: Yup, it's a number type

File: packages/mobile/src/identity/verification.ts
Comment (tarikbellamine) [reply packages/mobile/src/identity/verification.ts:76]: Yeah, I was thinking of scrapping this because we can just calculate the difference between the `verification_start` event and everything else. Decided to hold off on doing it until we have bandwidth to focus on creating more complex data queries but hope to eliminate these timing metrics sometime soon

File: packages/mobile/src/invite/saga.ts
Comment (tarikbellamine) [reply packages/mobile/src/invite/saga.ts]: Ah, gotcha. I will change the error message.

File: packages/mobile/src/app/saga.ts
Comment (tarikbellamine) [reply packages/mobile/src/app/saga.ts]: Good catch. Actually don't need this anymore because looks like segment gets the app state changes.

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [reply packages/mobile/src/analytics/ValoraAnalytics.ts:19]: Yeah good points here. The verification debugging bit is a strong argument. I guess I'm okay with keeping it included for a while.

File: packages/mobile/src/app/ErrorBoundary.tsx
Comment (tarikbellamine) [reply packages/mobile/src/app/ErrorBoundary.tsx:27]: What we are doing now is exclusively capturing device info on app load. Device info rarely changes so the thought is we should only really log it once per session at a maximum. From there, we can tie in device info for any particular event by doing a join query on sessionId. This minimizes payload sizes, which should have positive performance + network implications.

Do you foresee any issues with this approach?

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [reply packages/mobile/src/analytics/ValoraAnalytics.ts]: Yeah this is a tricky one. On the one hand, all on-chain activity is already publicly tied to an address. On the other, addresses are the blockchain-equivalent of user IDs so I feel like including it weakens our claim that we do analytics in an anonymous way. 

It certainly helps us make correlations across services (like identifying misbehaving PGPNP users). 

I have an inherent bias towards logging less, tracking less, but I'll defer to you three on this one :)

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [inline packages/mobile/src/analytics/ValoraAnalytics.ts]: This needs to go in init now right?

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [inline packages/mobile/src/analytics/ValoraAnalytics.ts]: Recommend making this async in possible and then using `await` instead of hooking on `.catch` and `.then`. callbacks

File: packages/mobile/src/alert/actions.ts
Comment (tarikbellamine) [reply packages/mobile/src/alert/actions.ts:69]: Oh my bad, I missed this use case and will add it back. Was just trying to cut down on repetitiveness because we are already logging errors with context specific error events, on error boundary, on showError, and also here.

Thinking about if there's a better way to do this for the next PR

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [inline packages/mobile/src/analytics/ValoraAnalytics.ts]: How about we call this from `index.js` or `App.tsx` for greater control of when it runs. By putting it here, it'll execute whenever this file happens to be first imported, and it'll block until init returns.

Comment (tarikbellamine) [reply packages/mobile/src/analytics/ValoraAnalytics.ts]: Got it. Put it in `App` right before I call `startSession`. I considered folding `init` and `startSession` into one method but figured it's better to keep the concerns a little more separate.

File: packages/mobile/src/analytics/ValoraAnalytics.ts
Comment (jmrossy) [inline packages/mobile/src/analytics/ValoraAnalytics.ts]: Nit, looks like we don't even need the constructor anymore at this point, you can just assign these defaults above.

**Review by jmrossy — APPROVED** [review]

**Pull request merged**
